RFCs/FS-1135-random-functions-for-collections.md

```diff
@@ -0,0 +1,206 @@
+# F# RFC FS-1135 - Random functions for collections (List, Array, Seq)
+
+The design suggestion [Add shuffle, sample etc. methods for lists, arrays etc](https://github.com/fsharp/fslang-suggestions/issues/508) has been marked "approved in principle".
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/508)
+- [x] Approved in principle
+- [ ] [Implementation]() (no implementation yet)
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- [Discussion](https://github.com/fsharp/fslang-design/discussions/731)
+
+# Summary
+
+This feature extends collections apis with functions for random sampling and shuffling built-in fsharp collections.
+
+# Motivation
+
+This feature is motivated by the following use cases:
+ - Using F# for data science and machine learning (like building a neural network), where data shuffling plays important role
+ - Building games, where random sampling is used for generating random levels, random decks, etc.
+ - Building simulations, where random sampling is used for generating random input data
+
+# Detailed design
+
+### General
+
+The following general rules are applied to all functions
+ - New functions should be implemented in List, Array, Seq modules
+ - All functions should have a variant with a [Random](https://learn.microsoft.com/en-us/dotnet/api/system.random) parameter
+ - Shared thread-safe Random instance should be used for all basic functions
+
+### Shuffle
+
+Shuffle function should returned a new shuffled collection of the same collection type.
+
+Two functions should be added to each module.
+
+```fsharp
+// Array module
+val randomShuffle: array:'T[] -> 'T[]
+val randomShuffleWith: random:Random -> array:'T[] -> 'T[]
+val randomShuffleInPlace: array:'T[] -> 'T[]
+val randomShuffleInPlaceWith: random:Random -> array:'T[] -> 'T[]
+// List module
+val randomShuffle: list:'T list -> 'T list
+val randomShuffleWith: random:Random -> list:'T list -> 'T list
+// Seq module
+val randomShuffle: source:'T seq -> 'T seq
+val randomShuffleWith: random:Random -> source:'T seq -> 'T seq
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.randomShuffle // [ "Charlie"; "Dave"; "Alice"; "Bob" ]
+```
+
+### Choice
+
+Choice function should returned a random element from the collection.
```

Comment:
nit:
```suggestion
The choice functions return a single random element from the given collection. The random choice is weighted evenly on the size of the collection.
```

Reply:
We might want to add a note about the direction of the implementation here. People may assume that it isn't necessary to traverse the whole sequence or list. However, since the choice should be balanced in relation to the size of the list or sequence, we will need to do a full traversal.

In the case of lists and sequences, that means that we (probably) need to create a temporary local copy of the full sequence. The alternative is multiple traversals.

Reply:
Fixed text. As for implementation guidance, I don't think it's necessary, since doing random selection on only part of collection will look unnatural and easy to spot on review.

Reply:
Good points, tx.